torch/csrc/jit/passes/onnx/fold_if_node.cpp

```diff
@@ -0,0 +1,249 @@
+
+#include <c10/util/Optional.h>
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+// This function determines wheather If Node can be folded.
+static bool isStaticCondition(Node* node) {
+  TORCH_INTERNAL_ASSERT(
+      node->kind() == onnx::If || node->kind() == onnx::Not ||
+      node->kind() == onnx::Identity);
+  auto cast_node = node->input()->node();
+  if (cast_node->kind() != onnx::Cast)
+    cast_node = node;
+  auto prev_node = cast_node->input()->node();
+
+  if (prev_node->kind() == onnx::Not || prev_node->kind() == onnx::Identity ||
+      prev_node->kind() == onnx::If)
+    return isStaticCondition(prev_node);
+
+  auto compare_node = prev_node;
+  if (compare_node->kind() == onnx::Equal ||
+      compare_node->kind() == onnx::Greater ||
+      compare_node->kind() == onnx::Less) {
+    for (size_t i = 0; i < compare_node->inputs().size(); i++) {
+      auto sym = compare_node->inputs()[i]
+                     ->type()
+                     ->cast<TensorType>()
+                     ->symbolic_sizes();
+      if (!(compare_node->inputs()[i]->node()->kind() == onnx::Constant ||
+            compare_node->inputs()[i]->node()->kind() == onnx::Size ||
+            compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd))
+        return false;
+      if (compare_node->inputs()[i]->node()->kind() != onnx::Constant) {
+        auto shape_node = compare_node->inputs()[i]->node()->input()->node();
+        auto shape =
+            shape_node->input()->type()->cast<TensorType>()->symbolic_sizes();
+
+        // ONNX shape and type inference cannot determine the shape of the input
+        if (!shape.rank())
+          return false;
+
+        // If dynamic_axes are used on inputs to ReduceProd node, don't fold If
+        // node
+        auto dynamic_axes = shape.isComplete();
+        if (!dynamic_axes &&
+            compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd)
+          return false;
+      }
+    }
+    return true;
+  } else if (compare_node->kind() == onnx::Constant) {
+    return true;
+  }
+  return false;
+}
+
+// This function returns the value of the constant-folded subblock
+// that is input to the If node.
+static bool constantFoldedConditionValue(Node* node) {
+  TORCH_INTERNAL_ASSERT(node->kind() == onnx::If);
+  // usually Cast node precedes If node in the graph, but
+  // there are some rare scenarios when that is not the case.
+  auto cast_node = node->input()->node();
+  if (cast_node->kind() != onnx::Cast)
+    cast_node = node;
+  auto prev_node = cast_node->input()->node();
+  if (prev_node->kind() == onnx::If) {
+    int cond = 1 - (int)constantFoldedConditionValue(prev_node);
+    Block* block = prev_node->blocks()[cond];
+    // we are assuming that the number of block outputs is 1.
+    // TODO: make this more general by supporting the case
+    // of multiple outputs
+    prev_node = block->outputs()[0]->node();
```

Comment:
You should be able to track back the precise input from `auto prev_node = cast_node->input()->node();`

the idx to use for `block->outputs()[idx]->node();` should match with the idx of `cast_node->input()` in `cast_node->input()->node()->outputs()`.

Reply:
done.